File: passenger-service/src/main/java/io/voitovich/yura/passengerservice/security/converter/JwtAuthConverter.java
package io.voitovich.yura.passengerservice.security.converter;

import org.springframework.core.convert.converter.Converter;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class JwtAuthConverter implements Converter<Jwt, AbstractAuthenticationToken> {

    @Override
    public AbstractAuthenticationToken convert(Jwt source) {
        Map<String, Object> claims = source.getClaims();
        Collection<GrantedAuthority> grantedAuthority = getGrantedAuthority(source);
        return new JwtAuthenticationToken(source, grantedAuthority);
    }

    public Collection<GrantedAuthority> getGrantedAuthority(Jwt source) {
        Map<String, Object> realmAccess = source.getClaimAsMap("realm_access");
        List<String> roles = (List<String>) realmAccess.get("roles");

        return roles.stream()
                .map(o -> "ROLE_" + o)
                .map(SimpleGrantedAuthority::new)
                .collect(Collectors.toList());
    }
}

File: ride-service/src/main/kotlin/io/voitovich/yura/rideservice/security/converter/JwtAuthConverter.kt
package io.voitovich.yura.rideservice.security.converter

import org.springframework.core.convert.converter.Converter
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import java.util.stream.Collectors

class JwtAuthConverter : Converter<Jwt, AbstractAuthenticationToken> {

    override fun convert(source: Jwt): AbstractAuthenticationToken {
        val claims = source.claims
        val grantedAuthority = getGrantedAuthority(source)
        return JwtAuthenticationToken(source, grantedAuthority)
    }

    private fun getGrantedAuthority(source: Jwt): Collection<GrantedAuthority> {
        val realmAccess = source.getClaimAsMap("realm_access")
        val roles = realmAccess["roles"] as List<String>?
        return roles!!.stream()
            .map { o: String -> "ROLE_$o" }
            .map { role: String? -> SimpleGrantedAuthority(role) }
            .collect(Collectors.toList())
    }
}